starlark/eval.go

```diff
@@ -836,14 +826,50 @@ unknown:
 	return nil, fmt.Errorf("unknown binary op: %s %s %s", x.Type(), op, y.Type())
 }
 
-func repeat(elems []Value, n int) (res []Value) {
-	if n > 0 {
-		res = make([]Value, 0, len(elems)*n)
-		for i := 0; i < n; i++ {
-			res = append(res, elems...)
-		}
+// It's always possible to overeat in small bites but we'll
+// try to stop someone swallowing the world in one gulp.
+const maxAlloc = 10000000
+
+func tupleRepeat(elems Tuple, n Int) (Tuple, error) {
+	if len(elems) == 0 {
+		return nil, nil
+	}
+	i, err := AsInt32(n)
+	if err != nil {
+		return nil, fmt.Errorf("repeat count %s too large", n)
+	}
+	if i < 1 {
+		return nil, err
+	}
+	// Inv: i > 0, len > 0
+	sz := len(elems) * i
```

Comment:
Is this actually 32-bit multiplication? `len(elems)` should be `int`, which would be 64-bit on`amd64`, so it would be 64-bit multiplication, right?

I think it would be clearer to check that `len(elems)` can be cast to a 32-bit integer without loss, then do the multiplication. You shouldn't need to check `sz < 0` after that.

Reply:
No, it's int multiplication, which is either 32 or 64. It can overflow on 64, though that's not realistic because i is < 2^32 which means the len would have to be larger than 2^32. I removed the "32" from the comment to make it more accurate.

I would rather not downgrade to 32 bits if the machine is 64, as is the norm. The <0 check is the usual way to detect multiplication overflow.

Reply:
I upped the limit to about a billion elements since that's large (by Starlark standards) but still falls into the range malloc can easily handle.